[NJT] Add cumsum support for nested tensors

ketansingh opened the pull request:
Summary: - Add cumsum support for NT

Test Plan: - added unit tests

Differential Revision: D68307097

ketansingh commented:
/easycla

jbschlosser commented on torch/nested/_internal/ops.py:
```diff
@@ -1007,6 +1007,42 @@ def chunk_default(func, *args, **kwargs):
+    inp = new_kwargs.pop("input")
+    dim = new_kwargs.pop("dim")
+
+    min_seqlen = inp._maybe_min_seqlen
+    max_seqlen = inp._maybe_max_seqlen
+    padded_max_S = max_seqlen
+
+    total_L = inp._values.shape[inp._ragged_idx - 1]
+    if padded_max_S is None:
+        # use upper bound on max seqlen if it's not present
+        padded_max_S = total_L
+
+    padded_shape = (
+        *inp.shape[: inp._ragged_idx],
+        padded_max_S,
+        *inp.shape[inp._ragged_idx + 1 :],
+    )
+    padded_inp = inp.to_padded_tensor(0.0, output_size=padded_shape)
+    return nested_from_padded(
+        func(padded_inp, dim),
+        offsets=inp._offsets,
+        ragged_idx=inp._ragged_idx,
+        sum_S=total_L,
+        min_seqlen=min_seqlen,
+        max_seqlen=max_seqlen,
+    )
```
I think we only need to do this padding-based approach if operating on the ragged dim: `dim == inp._ragged_idx`. We should error out nicely for the batch dim (`dim == 0`) and redispatch for other dims with the correct inner-space dim via `_wrap_jagged_dim()`.

ketansingh replied:
Thanks @jbschlosser I have mostly implemented your suggestions. Trying to figure out whether we need to support jagged dimensions in between for something like below (also, it results in error somehow)


```
tensor_shapes = [
    (2, 3, 4, 5, 6),  # First tensor with shape (2, 3, 4, 5, 6)
    (2, 3, 5, 5, 6),  # Second tensor with shape (2, 3, 5, 5, 6)
    (2, 3, 3, 5, 6),  # Third tensor with shape (2, 3, 3, 5, 6)
]
# Create a list of tensors with the specified shapes
tensors = [torch.randn(shape) for shape in tensor_shapes]

# Create a nested tensor from the list of tensors
nt = torch.nested.nested_tensor(tensors)

# Print the nested tensor
nt.shape

RuntimeError: Internal error: NestedTensorImpl doesn't support sizes. Please file an issue.

```

jbschlosser replied:
Agreed, I think it'd be ideal to support `cumsum()` on the ragged dim no matter where that dim is. We don't have to support this all at once, but we should be very explicit via errors for the cases we don't support initially.

> I was thinking for above example we only support dim>=3. WDYT?

Yep I think this is right. Smallest possible case is e.g. `(3, 6, j1)`, in which case we'd have a single dense dimension between the batch and ragged dims.

jbschlosser commented on test/test_nestedtensor.py:
```diff
@@ -1646,6 +1646,23 @@ def test_nested_tensor_mul_in_place(self, device, dtype):
 
+    @onlyCPU
+    @skipMeta
+    def test_nested_tensor_cumsum(self, device):
```
This test is good for the simple case, but in general we try to cover the cross product of `(eager, compile) x (forward, backward) x njt dim x (contig, non-contig with holes, non-contig transposed) x operating on (batch dim, ragged dim, other dim)`, etc.

To get this relatively easily, you can use the auto-generated OpInfo tests in `test/test_nestedtensor.py`. This will get you compile and eager tests for forward & backward run over a set of SampleInputs. SampleInputs for `cumsum()` can be defined in `torch/testing/_internal/opinfo/nested.py` and there are helpers there to make generating different types of NJTs easy. Lmk if you want more detail on this and I'm happy to help.

ketansingh committed:
[NJT] Add cumsum support for nested tensors (#145903)

Summary:

- Add cumsum support for NT

Test Plan: - added unit tests

Differential Revision: D68307097
